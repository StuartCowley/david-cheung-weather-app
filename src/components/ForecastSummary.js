import React from "react";
import PropTypes from "prop-types";
import WeatherIcon from "react-icons-weather";
import "../styles/forecast-summary.css";

function dateOrdinal(d) {
  let subfix;
  if (d === "31" || d === "21" || d === "01") {
    subfix = "st";
  } else if (d === "22" || d === "02") {
    subfix = "nd";
  } else if (d === "23" || d === "03") {
    subfix = "rd";
  } else {
    subfix = "th";
  }
  if (d.charAt(0) === "0") {
    return d.charAt(1) + subfix;
  }
  return d + subfix;
}

function ForecastSummary(props) {
  const { date, temperature, description, icon, onSelect } = props;
  const tempDate1 = new Date(date).toDateString(); // Mon Apr 30 2018
  const tempDate2 = tempDate1.split(" ");
  const formattedDate = `${tempDate2[0]} ${dateOrdinal(tempDate2[2])} ${
    tempDate2[1]
  }`; // Man 30th Apr

  return (
    <div className="forecast-summary" data-testid="forecast-summary">
      <div className="forecast-summary__date">{formattedDate}</div>
      <div className="forecast-summary__icon" data-testid="forecast-icon">
        <WeatherIcon name="owm" iconId={icon.toString()} />
      </div>
      <div className="forecast-summary__description">{description}</div>
      <div className="forecast-summary__temperature">
        {temperature.max}
        &deg;C
      </div>

      <button
        type="button"
        onClick={() => {
          onSelect(date);
        }}
      >
        More details
      </button>
    </div>
  );
}

export default ForecastSummary;

ForecastSummary.propTypes = {
  date: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  icon: PropTypes.number.isRequired,
  temperature: PropTypes.shape({
    min: PropTypes.number,
    max: PropTypes.number,
  }).isRequired,
  // other props validation
  onSelect: PropTypes.func.isRequired,
};
