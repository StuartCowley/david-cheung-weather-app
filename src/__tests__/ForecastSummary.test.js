import React from "react";
import { render, screen } from "@testing-library/react";
import ForecastSummary from "../components/ForecastSummary";
import renderer from "react-test-renderer";

function decode(str) {
  let txt = document.createElement("textarea");
  txt.innerHTML = str;
  return txt.value;
}

describe("ForecastSummary", () => {
  const validProps = {
    date: 1525046400000,
    description: "Stub description",
    icon: 200,
    temperature: {
      min: 12,
      max: 22,
    },
    onSelect: () => {},
  };

  test("Renders as expected", () => {
    const rendered = renderer.create(
      <ForecastSummary
        date={validProps.date}
        temperature={validProps.temperature}
        description={validProps.description}
        icon={validProps.icon}
        onSelect={validProps.onSelect}
      />
    );
    expect(rendered).toMatchSnapshot();
  });

  test("Assert all classname elements are present", () => {
    render(
      <ForecastSummary
        date={validProps.date}
        temperature={validProps.temperature}
        description={validProps.description}
        icon={validProps.icon}
        onSelect={validProps.onSelect}
      />
    );
    expect(document.querySelector(".forecast-summary")).toBeTruthy();
    expect(document.querySelector(".forecast-summary__date")).toBeTruthy();
    expect(document.querySelector(".forecast-summary__icon")).toBeTruthy();
    expect(
      document.querySelector(".forecast-summary__temperature")
    ).toBeTruthy();
    expect(
      document.querySelector(".forecast-summary__description")
    ).toBeTruthy();
  });

  test("Assert all props elements are present", () => {
    render(
      <ForecastSummary
        date={validProps.date}
        temperature={validProps.temperature}
        description={validProps.description}
        icon={validProps.icon}
        onSelect={validProps.onSelect}
      />
    );
    expect(screen.getByText("Mon 30th Apr")).toBeTruthy();
    expect(screen.getByText(`22${decode("&deg;C")}`)).toBeTruthy();
    expect(screen.getByText("Stub description")).toBeTruthy();
    // expect(getByTestId("forecast-icon")).toHaveTextContent("200");

    expect(screen.getByText("Mon 30th Apr")).toBeInstanceOf(HTMLDivElement);
    expect(screen.getByText(`22${decode("&deg;C")}`)).toBeInstanceOf(
      HTMLDivElement
    );
    expect(screen.getByText("Stub description")).toBeInstanceOf(HTMLDivElement);
    expect(screen.getByTestId("forecast-icon")).toBeInstanceOf(HTMLDivElement);
  });
});
