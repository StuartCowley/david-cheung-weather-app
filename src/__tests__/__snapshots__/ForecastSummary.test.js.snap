// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastSummary Renders as expected 1`] = `
<div
  className="forecast-summary"
  data-testid="forecast-summary"
>
  <div
    className="forecast-summary__date"
  >
    Mon 30th Apr
  </div>
  <div
    className="forecast-summary__icon"
    data-testid="forecast-icon"
  >
    <i
      className="wi wi-day-storm-showers"
      name="owm"
    />
  </div>
  <div
    className="forecast-summary__description"
  >
    Stub description
  </div>
  <div
    className="forecast-summary__temperature"
  >
    22
    Â°C
  </div>
  <button
    onClick={[Function]}
    type="button"
  >
    More details
  </button>
</div>
`;
